internal/rbd/manager.go

```diff
@@ -268,3 +316,108 @@ func (mgr *rbdManager) CreateVolumeGroup(ctx context.Context, name string) (type
+	pool, err := vg.GetPool(ctx)
+	if err != nil {
+		return nil, err
+	}
+
+	// volumeNamePrefix is an optional parameter, can be an empty string
+	prefix := mgr.parameters["volumeNamePrefix"]
+
+	// journalPool is an optional parameter, use pool if it is not set
+	journalPool, ok := mgr.parameters["journalPool"]
+	if !ok || journalPool == "" {
+		journalPool = pool
+	}
+
+	clusterID, err := vg.GetClusterID(ctx)
+	if err != nil {
+		return nil, fmt.Errorf("failed to get cluster id for volume group snapshot %q: %w", vg, err)
+	}
+
+	uuid, freeUUID, err := mgr.getGroupUUID(ctx, clusterID, journalPool, name, prefix)
+	defer func() {
+		// no error, no need to undo the reservation
+		if err == nil {
+			return
+		}
+
+		freeUUID()
+	}()
+
+	monitors, err := util.Mons(util.CsiConfigFile, clusterID)
+	if err != nil {
+		return nil, fmt.Errorf("failed to find MONs for cluster %q: %w", clusterID, err)
+	}
+
+	_ /*journalPoolID*/, poolID, err := util.GetPoolIDs(ctx, monitors, journalPool, pool, mgr.creds)
+	if err != nil {
+		return nil, fmt.Errorf("failed to get PoolID for JournalPool %s and ImagePool %s: %w", journalPool, pool, err)
+	}
+
+	csiID, err := util.GenerateVolID(ctx, monitors, mgr.creds, poolID, pool, clusterID, uuid)
+	if err != nil {
+		return nil, fmt.Errorf("failed to generate a unique CSI volume group with uuid for %q: %w", uuid, err)
+	}
+
+	vgs, err := rbd_group.GetVolumeGroupSnapshot(ctx, csiID, mgr.csiID, mgr.creds, mgr)
+	if vgs != nil {
+		log.DebugLog(ctx, "found existing volume group snapshot %q for id %q", vgs, csiID)
+
+		// validate the contents of the vgs
+		snapshots, vgsErr := vgs.ListSnapshots(ctx)
+		if vgsErr != nil {
+			return nil, fmt.Errorf("failed to list snapshots of existing volume group snapshot %q: %w", vgs, vgsErr)
+		}
+
+		volumes, vgErr := vg.ListVolumes(ctx)
+		if vgErr != nil {
+			return nil, fmt.Errorf("failed to list volumes of volume group %q: %w", vg, vgErr)
+		}
+
+		// contents matches, return the existing vgs
+		if len(snapshots) == len(volumes) {
+			log.DebugLog(ctx, "existing volume group snapshot %q contains %d snapshots", vgs, len(snapshots))
+
+			return vgs, nil
+		}
+	} else if err != nil && !errors.Is(ErrImageNotFound, err) {
+		return nil, fmt.Errorf("failed to check for existing volume group snapshot with id %q: %w", csiID, err)
+	}
+
+	snapshots, err := vg.CreateSnapshots(ctx, mgr.creds, csiID)
+	if err != nil {
+		return nil, fmt.Errorf("failed to create volume group snapshot %q: %w", name, err)
+	}
+	defer func() {
+		// cleanup created snapshots in case there was a failure
+		if err == nil {
+			return
+		}
+
+		for _, snap := range snapshots {
+			delErr := snap.Delete(ctx)
+			if delErr != nil {
+				log.ErrorLog(ctx, "failed to delete snapshot %q: %v", snap, delErr)
+			}
+		}
+	}()
+
+	log.DebugLog(ctx, "volume group snapshot %q contains %d snapshots: %v", name, len(snapshots), snapshots)
+
+	vgs, err = rbd_group.NewVolumeGroupSnapshot(ctx, csiID, mgr.csiID, mgr.creds, snapshots)
```

Comment:
we are passing the same csiID twice here, is this fine?

Reply:
https://github.com/ceph/ceph-csi/blob/f31e3917c209ff3d00c8c3b3ab5f3d8181db57a4/internal/rbd/group/group_snapshot.go#L113-L122 

one is volgroupsnap ID/handle and other is csidriver instance id

Reply:
> we are passing the same csiID twice here, is this fine?

As Rakshith points out, these are not the same. One is the ID for the new VGS, the other is the ID of the driver.

Reply:
got confused because of using same variable name for both,  can user rename `csiID` to `groupID` or `id` to avoid confusion